File: Study Plan/Top Interview 150/0121 Best Time to Buy and Sell Stock/BestTimeToBuyAndSellStock.java
/*  
    Problem
    --------------------------------------------------------------------------------
    121. Best Time to Buy and Sell Stock
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/best-time-to-buy-and-sell-stock/solutions/4447173/java-time-complexity-o-n-space-complexity-o-1-beats-75-in-runtime/
    
    - Time Complexity O(n)
    - Space Complexity O(1)

    - Runtime 2 ms (Beats 75.39% of users with Java)
    - Memory 61.74 MB (Beats 9.44% of users with Java)
    
*/

class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0;
        int buyingPrice = prices[0];
        for (int day = 1; day < prices.length; day++) {
            int dailyPrice = prices[day];
            if (dailyPrice < buyingPrice) {
                // buy cheaper
                buyingPrice = dailyPrice;
            } else if (dailyPrice - buyingPrice > profit) {
                // sell and take profit
                profit = dailyPrice - buyingPrice;
            }
        }
        return profit;
    }
}

File: Study Plan/Top Interview 150/0121 Best Time to Buy and Sell Stock/BestTimeToBuyAndSellStock.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    121. Best Time to Buy and Sell Stock
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/best-time-to-buy-and-sell-stock/solutions/4384292/c-time-complexity-o-n-space-complexity-o-1-beats-72-in-runtime/

    - Time Complexity O(n)
    - Space Complexity O(1)

    - Runtime 207 ms (Beats 72.14% of users with C#)
    - Memory 49.80 MB (Beats 78.80% of users with C#)
*/
public class Solution
{
    public int MaxProfit(int[] prices)
    {
        var profit = 0;
        var buyingPrice = prices[0];

        for (var day = 1; day < prices.Length; day++)
        {
            var dailyPrice = prices[day];

            // https://github.com/LeetCode-Feedback/LeetCode-Feedback/issues/18237
            // if (dailyPrice > 0)
            // {
                if (dailyPrice < buyingPrice)
                {
                    // buy cheaper
                    buyingPrice = dailyPrice;
                }
                else if (dailyPrice - buyingPrice > profit)
                {
                    // sell and take profit
                    profit = dailyPrice - buyingPrice;
                }
            // }
        }

        return profit;
   